grin/src/stratumserver.rs

```diff
@@ -0,0 +1,518 @@
+	height: u64,
+	nonce: u64,
+	pow: Vec<u32>,
+}
+
+// ----------------------------------------
+// Worker Factory Thread Function
+
+// Run in a thread. Adds new connections to the workers list
+fn accept_workers(id: String, address: String, workers: &mut Arc<Mutex<Vec<Worker>>>) {
+	let listener = TcpListener::bind(address).expect("Failed to bind to listen address");
+	let mut worker_id: u32 = 0;
+	for stream in listener.incoming() {
+		match stream {
+			Ok(stream) => {
+				warn!(
+					LOGGER,
+					"(Server ID: {}) New connection: {}",
+					id,
+					stream.peer_addr().unwrap()
+				);
+				stream
+					.set_nonblocking(true)
+					.expect("set_nonblocking call failed");
+				let mut worker = Worker::new(worker_id.to_string(), BufStream::new(stream));
+				workers.lock().unwrap().push(worker);
+				worker_id = worker_id + 1;
+			}
+			Err(e) => {
+				warn!(
+					LOGGER,
+					"(Server ID: {}) Error accepting connection: {:?}", id, e
+				);
+			}
+		}
+	}
+	// close the socket server
+	drop(listener);
+}
+
+// ----------------------------------------
+// Worker Object - a connected stratum client - a miner, pool, proxy, etc...
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct JobTemplate {
+	pre_nonce: String,
+	post_nonce: String,
+}
+
+pub struct Worker {
+	id: String,
+	stream: BufStream<TcpStream>,
+	error: bool,
+	authenticated: bool,
+}
+
+impl Worker {
+	/// Creates a new Stratum Worker.
+	pub fn new(id: String, stream: BufStream<TcpStream>) -> Worker {
+		Worker {
+			id: id,
+			stream: stream,
+			error: false,
+			authenticated: false,
+		}
+	}
+
+	// Get Message from the worker
+	fn read_message(&mut self) -> Option<String> {
+		// Read and return a single message or None
+		let mut line = String::new();
+		match self.stream.read_line(&mut line) {
+			Ok(_) => {
+				return Some(line);
+			}
+			Err(ref e) if e.kind() == ErrorKind::WouldBlock => {
+				// Not an error, just no messages ready
+				return None;
+			}
+			Err(e) => {
+				warn!(
+					LOGGER,
+					"(Server ID: {}) Error in connection with stratum client: {}", self.id, e
+				);
+				self.error = true;
+				return None;
+			}
+		}
+	}
+
+	// Send Message to the worker
+	fn send_message(&mut self, message_in: String) {
+		// Write and Flush the message
+		let mut message = message_in.clone();
+		if !message.ends_with("\n") {
+			message += "\n";
+		}
+		match self.stream.write(message.as_bytes()) {
+			Ok(_) => match self.stream.flush() {
+				Ok(_) => {}
+				Err(e) => {
+					warn!(
+						LOGGER,
+						"(Server ID: {}) Error in connection with stratum client: {}", self.id, e
+					);
+					self.error = true;
+				}
+			},
+			Err(e) => {
+				warn!(
+					LOGGER,
+					"(Server ID: {}) Error in connection with stratum client: {}", self.id, e
+				);
+				self.error = true;
+				return;
+			}
+		}
+	}
+} // impl Worker
+
+// ----------------------------------------
+// Grin Stratum Server
+
+pub struct StratumServer {
+	id: String,
+	miner: Miner,
```

Comment:
The stratum server shouldn't need the miner, just the pool, the chain and the utility to build a block. We should be able at some point, if we decide to, to remove the embedded miner and keep the stratum server.